### ISSUE TITLE
Exchange federation link worker sup IDs not handled properly after upgrade to 3.13

### ISSUE THREAD

**Issue by gomoripeti** [rabbitmq/rabbitmq-server#10306]
### Describe the bug

Tested in 3.13.0-rc.3 (but I don't see any change in the federation plugin later in 3.13.0-rc.4)

During an upgrade (multi-node cluster with rolling restart) from pre 3.13.0 version with existing federation exchange links, the mirrored supervisor preserves old child id format. This can cause problems on the new version when the exchange worker should be stopped - when updating the policy or trying to delete a federated exchange.

- Trying to delete a federated exchange results in the below crash (and the exchange is not deleted)
```
Request to node rabbit@hostname failed with {'EXIT',
{{badmatch,
{error,
not_found}},
[{rabbit_federation_exchange_link_sup_sup,
stop_child,1,
[{file,
"rabbit_federation_exchange_link_sup_sup.erl"},
{line,104}]},
{rabbit_federation_exchange,
maybe_stop,1,
[{file,
"rabbit_federation_exchange.erl"},
{line,93}]},
{rabbit_exchange,
'-callback0/5-lc$^2/1-2-',
5,
[{file,
"rabbit_exchange.erl"},
{line,65}]},
{rabbit_exchange,
callback0,5,
[{file,
"rabbit_exchange.erl"},
{line,65}]},
{rabbit_exchange,
callback,4,
[{file,
"rabbit_exchange.erl"},
{line,55}]},
{rabbit_binding,
'-process_deletions/1-fun-0-',
2,
[{file,
"rabbit_binding.erl"},
{line,431}]},
```

I see two ways to fix this, and appreciate feedback on which way to go

- Revert to old id format if khepri is disabled

  During migration to Khepri old child ids are converted to new
  format anyway. So let's revert and use old child id format also on
  3.13.0 if Mnesia is used. This is to avoid the need to handle
  two formats at the same time.

  This is simpler but only acceptable if there is no strict
  requirement to support rolling upgrade from 3.13.0-rc.x to 3.13.0

  [WIP commit](https://github.com/cloudamqp/rabbitmq-server/commit/87716efa8e91cd88d149b8c5b5860d9f4af75279)

- Handle old child id format (similar to shovel dynamic workers)

  Requires more compatibility code

  [WIP commit](https://github.com/cloudamqp/rabbitmq-server/commit/bc07ee1d51435f6f7c7ca5c780e02285179d1dbc)

Fortunately federated queues seem to be unaffected because of the
difference in exchange and queue decorator behaviours. In case of
queues the federation link worker runs on the same node as the queue
leader and is shut down and deleted when the queue is shut down (ie at
a graceful node shutdown). During a rolling upgrade this shutdown uses
the old code which is compatible with the old child id format. Then
the worker is added on the new node where the new queue leader is
started. So the child id format stays compatible with the node's
version it runs on.

### Reproduction steps

1. Create a 3-node cluster with version 3.12.x (that will be upgraded, let's call this SUT) and another 1-node cluster (serving as federation upstream)
2. Create a federation upstream on SUT called "fed_on_upgrade" with parameters
```
  "parameters": [
    {
      "value": {
        "ack-mode": "on-confirm",
        "trust-user-id": false,
        "uri": "amqps://<url of upstream cluster>"
      },
      "component": "federation-upstream",
      "name": "fed_on_upgrade"
    }
  ]
```
3. Create an exchange and a policy on SUT eg by uploading the bellow definition
```
{
  "policies": [
    {
      "name": "fed_on_upgrade_policy",
      "pattern": "^fed_.*",
      "apply-to": "all",
      "definition": {
        "federation-upstream": "fed_on_upgrade"
      },
      "priority": 2
    }
  ],
  "exchanges": [
    {
      "name": "fed_ex_on_upgrade_cluster",
      "type": "direct",
      "durable": true,
      "auto_delete": false,
      "internal": false,
      "arguments": {}
    }
  ]
}
```
4.  On any node on SUT in a remote_shell verify expected supervisor child
```
> r(rabbit_exchange).
> [{#exchange{name = #resource{name = <<"fed_ex_on_upgrade_cluster">>}}, _, _, _}] = mirrored_supervisor:which_children(rabbit_federation_exchange_link_sup_sup), ok.
```
5. Execute a rolling upgrade of SUT to 3.13.0-rc.4
6. Observe that supervisor child id did not change by executing point 4. again
7. In the Management UI of SUT try to delete exchange "fed_ex_on_upgrade_cluster" -> the UI popup shows a `not_found` crash stacktrace and the exchange is not deleted.

### Expected behavior

Federated exchanges (and exchange federation links) created before an upgrade to 3.13.0 should fully work after an upgrade to 3.13.0

**Comment by michaelklishin** [comment]
@gomoripeti can't we do exactly what we did to Shovels? It sounds like the same problem in a different place to me.

**Comment by gomoripeti** [comment]
@michaelklishin indeed this is the exact same problem. But this is not as bad because the id format change was not backported for federation, so all already released versions (except release candidates) still use the old id format. So there is a simpler solution (which is enabled by https://github.com/rabbitmq/rabbitmq-server/pull/10096): "Revert to old id format if khepri is disabled." My reason to prefer this one is because it took about 3 patch releases to get shovels right.

**Comment by michaelklishin** [comment]
@gomoripeti nonetheless, we did settle on a certain solution for Shovel. Mnesia ("Khepri is not enabled") will go extinct in the land of RabbitMQ in a single digit number of months.

The hard part is the upgrade path, and it cannot be avoided.

**Comment by gomoripeti** [comment]
sorry for the delayed response, finally I had time to spend on this topic

I agree that the upgrade path is hard. I see it like this

3.12 Mnesia -(step 1.)-> 3.13 Mnesia -(step 2.)-> 3.13 Khepri

- 3.12 Mnesa only supports old child id format
- 3.13 Mnesia only supported new child id format and there is no conversion in step 1.
  after lot of testing and trial and error I changed my mind about the fix and submitting a PR to map the solution implemented for "shovels"
- 3.13 Khepri only supports new child id format

- in step 2. a record conversion from old child id format to new one was implemented in #10096 
However during testing I found out that this is not complete, the supervisor state (in the process) still has the old child ids and this inconsistency leads to various failures after migrating to Khepri
I will submit a separate Issue to report this. Based on previous weeks I cannot commit to working on a fix for this unfortunately.

**Comment by michaelklishin** [comment]
Going straight from 3.12 to 3.13 with Khepri is possibly in theory but very unlikely for production systems, and we can document this fact.

I'm not sure what you mean by "based on previous weeks".

**Comment by michaelklishin** [comment]
Federation links are not that different from dynamic Shovels:

 * They are started at node boot
 * They have IDs and attach to a mirrored supervisor
 * They store some of their state in the schema data store

Why specifically can't the solution we have for shovels work here?

Also, can you be more specific than the convertion is "not complete"? Claims like that do not help us ship 3.13 sooner in any way, so we either do it in a way that would be problematic for CloudAMQP, or CloudAMQP will miss out on this RabbitMQ version and then 4.0, and all the fundamental improvements that come with not using Mnesia at all in RabbitMQ.

**Comment by michaelklishin** [comment]
@gomoripeti I will be blunt: we need more specifics here and either decide to address it, or 3.13.0 will ship as is, and it will be up to the companies that host RabbitMQ as a service to find a solution.

3.13 cannot be delayed for a few more months.

**Comment by gomoripeti** [comment]
sure, understandable and understood

I just wanted to make a sign-of-life. I will describe the "incomplete record migration" in detail and reproducible way. And it is also totally fine if this edge case does not block release of 3.13.0 and slowly improved in followup patch versions (either by me or by others)

**Comment by gomoripeti** [comment]
I opened https://github.com/rabbitmq/rabbitmq-server/issues/10440 to describe the "incomplete record migration" during Khepri migration issue.
This issue could track the 3.12 -> "3.13 Mnesia" upgrade part only, which should be addressed by PR #10416 .

**Comment by michaelklishin** [comment]
The 3.12 -> 3.13 Mnesia migration should be handled by #10453. See #10440 for future work related to Khepri and Federation, Shovel supervisor child IDs.

**Comment by dumbbell** [comment]
Just to clarify a few points, which should make a solution easier to find:

> Mnesia ("Khepri is not enabled") will go extinct in the land of RabbitMQ in a single digit number of months.

Very unlikely as it would be removed from RabbitMQ 4.1.0 at the earliest.

> Going straight from 3.12 to 3.13 with Khepri is possibly in theory

No, it's not possible: you have to upgrade, then enable `khepri_db`. To enable `khepri_db`, all nodes in the cluster must be running and know about `khepri_db`, like other feature flags.

I'm going to look at other issues and pull requests linked in this one.

**Comment by dumbbell** [comment]
I have a local prototype that approaches the problem differently: instead of changing the ID format and breaking running supervision trees and causing issues during upgrade, the ID is left alone but is converted to a Khepri-compatible path only when we need it.

This relies on the fact that the `Group` argument is always a module AFAICT, even if `mirrored_supervisor` documents that it can be any term.

Diana is in holidays currently, so I can't run the idea by her for now. I will prepare a more complete patch and publish it to GitHub so you can take a look.

**Comment by dumbbell** [comment]
Here is the branch:
https://github.com/rabbitmq/rabbitmq-server/tree/restore-mirrored_supervisor-child-id-format

What do you think @gomoripeti?

**Comment by lukebakken** [comment]
> the ID is left alone but is converted to a Khepri-compatible path only when we need it

👍 👍 👍

**Comment by lukebakken** [comment]
I am re-opening based on @dumbbell's [comment](https://github.com/rabbitmq/rabbitmq-server/issues/10306#issuecomment-1921568658)

**Comment by gomoripeti** [comment]
Thanks all for looking at this.
- On the administrative side this issue was addressing only the 3.13 on Mnesia case, and there was a PR from Michael merged addressing this. That's why this issue was closed. There is https://github.com/rabbitmq/rabbitmq-server/issues/10440 which tracks the Khepri path and friends problem.

- > the ID is left alone

  This is fine for federation but shovel plugin has the same issue. And there the new id format is already released in 3.12 and even 3.11. So there _might_ be deployments out there which already have running shovel workers with both old and new child id format. Then we have to think about how to convert the new id format back to the old one.

- Given the current situation that versions supporting both old and new child id format on Mnesia are already merged for federation and already released for shovels, what do you all think about a solution:
  Restart all mirrored supervisor workers at the start of khepri migration (from the init_copy_to_khepri callback) before copying of records starts. Either by using the assumption that `group` is a module or by a registration mechanism that Diana implemented.

**Comment by dumbbell** [comment]
The more I think about the problem, the more I believe we shouldn't use the new ID format. It duplicates data for no benefit. And while looking into this, I also discovered that the new ID format is already out there for shovels in 3.12.x and 3.11.x as you said.

We put so much effort into making sure the upgrade to Khepri was smooth for the past few years (and there are still rough edges as we can see), that I refuse to add code to federation to "fix" something that shouldn't be there in the first place.

The only acceptable solution to me is to find something to convert new IDs back to their original format in the shovel plugin.

**Comment by dumbbell** [comment]
The work in progress to address that upgrade problem is tracked in pull request #10472.

**Comment by dumbbell** [comment]
Fixed by #10472.
